### PR TITLE
[FIX] Supporting multi chunk for large documents

### PR DESCRIPTION
**Pull request by harini-venkataraman**
## What
This PR enables the following.
1. Support for proper response when large document is executued.

...
## Why
Larger docuemnts context retrival was not optimized for non zero chunks.

...

## How
Supporting Simple and Sub query retrieval.

...
## Can this PR break any existing features. If yes, please list possible items. If no, please explain why. (PS: Admins do not merge the PR without this section filled)

No, a method refactored in prompt service enables the support.
...

## Database Migrations
Added a field to Choices in model.
...

## Env Config
Not Applicable.
...

## Relevant Docs

...

## Related Issues or PRs

...

## Dependencies Versions

...

## Notes on Testing

Reproducing the error -
The response was proper for docuemtns with pages - 500,
but the ambiguity in response happend only after pages greater that 1000.

...

## Screenshots
Error reproduced 
Screenshot from 2024-05-13 22-31-22

Fix
Screenshot from 2024-05-14 14-41-53

Screenshot from 2024-05-13 18-13-45

## Checklist

I have read and understood the [Contribution Guidelines]().

### CODE REVIEW

File: prompt-service/src/unstract/prompt_service/main.py [hunk 1]
```diff
@@ -641,30 +641,48 @@ def prompt_processor() -> Any:
     return structured_output
 
 
-def simple_retriver(  # type:ignore
+def simple_or_subq_retriever(  # type:ignore
```
Comment (hari-kuriakose) [inline prompt-service/src/unstract/prompt_service/main.py:644]: @harini-venkataraman Let's rename this fn to `run_retrieval()`.

File: prompt-service/src/unstract/prompt_service/main.py [hunk 2]
```diff
@@ -641,30 +641,48 @@ def prompt_processor() -> Any:
...
     output: dict[str, Any],
     doc_id: str,
     llm_helper: ToolLLM,
     llm_li: Optional[LLM],
     vector_index,
+    retrival_type: str,
```
Comment (hari-kuriakose) [inline prompt-service/src/unstract/prompt_service/main.py:650]: @harini-venkataraman Let's fix typo and rename all variables to `retrieval_*`, etc.

File: prompt-service/src/unstract/prompt_service/main.py [hunk 3]
```diff
@@ -641,30 +641,48 @@ def prompt_processor() -> Any:
...
+        output,
+        llm_helper,
+        llm_li,
+        text,
+        "promptx",
+    )
+    return (answer, text)
+
+
+def generate_context(output, doc_id, vector_index, answer) -> str:
```
Comment (hari-kuriakose) [inline prompt-service/src/unstract/prompt_service/main.py:685]: @harini-venkataraman Let's rename this fn to `_retrieve_context()`.

File: prompt-service/src/unstract/prompt_service/main.py [hunk 4]
```diff
@@ -641,30 +641,48 @@ def prompt_processor() -> Any:
...
-        f" help extract relevant documents from a vector store:\n\n{prompt}"
-    )
+    context_prompt = ""
+    if retrival_type is PSKeys.SIMPLE:
+        context_prompt = prompt
+    if retrival_type is PSKeys.SUBQUESTION:
+        context_prompt = (
+            f"Generate a sub-question from the following verbose prompt that will"
+            f" help extract relevant documents from a vector store:\n\n{prompt}"
+        )
```
Comment (hari-kuriakose) [inline prompt-service/src/unstract/prompt_service/main.py:665]: @harini-venkataraman We need to invoke LLM via `run_completion()` only for subquestion retriever. For the simple retriever, we proceed with original prompt as is.

```suggestion
    if retrival_type is PSKeys.SUBQUESTION:
        subq_prompt = (
            f"Generate a sub-question from the following verbose prompt that will"
            f" help extract relevant documents from a vector store:\n\n{prompt}"
        )
        prompt = run_completion(
            llm_helper=llm_helper,
            llm_li=llm_li,
            prompt=context_prompt,
            adapter_instance_id=output[PSKeys.LLM],
        )

    context = _retrieve_context(output, doc_id, vector_index, prompt)

    answer = construct_and_run_prompt(  # type:ignore
        output,
        llm_helper,
        llm_li,
        text,
        "promptx",
    )

    return (answer, context)
```

**Review by hari-kuriakose — CHANGES_REQUESTED** [review]
@harini-venkataraman LGTM overall. Let's address the comments.

File: prompt-service/src/unstract/prompt_service/main.py (hunk 1 above)
Comment (harini-venkataraman) [reply prompt-service/src/unstract/prompt_service/main.py:644]: @hari-kuriakose I had the same dilemma, of making the function name generic to all retrievals or keeping it specific. When other retrieval methods are introduced, felt this naming will be explanatory. Your thoughts?

Comment (hari-kuriakose) [reply prompt-service/src/unstract/prompt_service/main.py:644]: @harini-venkataraman The function name would still be `run_retrieval()`.
And we would add more retrieval strategies to the same if else ladder.

If the logic for a particular retrieval strategy becomes complex, then we could branch out that alone to a private member function, that's all. Th entrypoint for all retrieval strategies would continue to be `run_retrieval()`.

**Review by hari-kuriakose — APPROVED** [review]
@harini-venkataraman LGTM

**Review by nehabagdia — APPROVED** [review]

**Pull request merged**